servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/customtransport/InMemoryServerTransport.java

```diff
@@ -0,0 +1,317 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.grpc.customtransport;
+
+import io.servicetalk.buffer.api.Buffer;
+import io.servicetalk.buffer.api.BufferAllocator;
+import io.servicetalk.concurrent.BlockingIterable;
+import io.servicetalk.concurrent.api.Publisher;
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.concurrent.api.internal.ConnectablePayloadWriter;
+import io.servicetalk.grpc.api.GrpcBindableService;
+import io.servicetalk.grpc.api.GrpcExecutionContext;
+import io.servicetalk.grpc.api.GrpcPayloadWriter;
+import io.servicetalk.grpc.api.GrpcServiceContext;
+import io.servicetalk.grpc.api.GrpcStatusException;
+import io.servicetalk.grpc.api.MethodDescriptor;
+import io.servicetalk.grpc.api.ParameterDescriptor;
+import io.servicetalk.grpc.api.SerializerDescriptor;
+import io.servicetalk.grpc.customtransport.Utils.ChannelGrpcServiceContext;
+import io.servicetalk.grpc.customtransport.Utils.GrpcStreamingDeserializer;
+import io.servicetalk.grpc.customtransport.Utils.GrpcStreamingSerializer;
+import io.servicetalk.grpc.customtransport.Utils.UtilGrpcExecutionContext;
+import io.servicetalk.serializer.api.StreamingDeserializer;
+import io.servicetalk.serializer.api.StreamingSerializer;
+import io.servicetalk.transport.netty.internal.GlobalExecutionContext;
+import io.servicetalk.transport.netty.internal.NettyIoExecutors;
+
+import com.google.rpc.Status;
+import io.netty.channel.Channel;
+import io.netty.channel.EventLoop;
+import io.netty.util.Attribute;
+import io.netty.util.AttributeKey;
+
+import java.io.IOException;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodType;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.concurrent.api.Publisher.defer;
+import static io.servicetalk.concurrent.api.Publisher.failed;
+import static io.servicetalk.concurrent.api.Publisher.from;
+import static io.servicetalk.grpc.api.GrpcStatusCode.INTERNAL;
+import static io.servicetalk.grpc.api.GrpcStatusCode.UNIMPLEMENTED;
+import static io.servicetalk.utils.internal.PlatformDependent.throwException;
+import static java.lang.Math.ceil;
+import static java.lang.invoke.MethodHandles.lookup;
+import static java.lang.invoke.MethodType.methodType;
+import static java.util.Objects.requireNonNull;
+
+final class InMemoryServerTransport implements ServerTransport {
```

Comment:
Is this only useful for test? What are the barriers to others using it besides APNs which has their own.

Reply:
The "custom transport" is not a common use case however there are some folks who need this. I would prefer to start with this as just a test to confirm it satisfies the requirements before making it more broadly visible. I'm concerned if we expose this in an example or as a standalone utility we will confuse folks looking for "normal gRPC", and also increase our API surface area more than necessary (if it continues to be a niche use case).

Reply:
I know we have the extensions project as well but this is the kind of thing that would normally go in a "contrib" area of open source projects; not core but potentially useful for specific cases or as sample code.